settings: Allow "resolve topic" permissions to be managed independently.

roanster007 opened the pull request:
This commit separates the "resolve topic" permissions from the topic editing permissions, through the introduction of two settings - "allow_resolve_topic" to control whether to allow resolve topics in organization, and "resolve_topic_policy" to restrict the resolve topics permissions.

In addition, it also restricts resolve topic permissions to streams where users have posting permissions.

Fixes #21811, #21281

[#api desing](https://chat.zulip.org/#narrow/stream/378-api-design/topic/managing.20resolve.20topic.20permissions.20independently/near/1795635) thread

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![image](https://github.com/user-attachments/assets/d290c029-c5db-460c-a58c-29493e1d8e19)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
So I guess I should have made sure folks were aware of this previously, but all future permissions settings need to be done using groups (see #19525), and likely building on the approach in https://github.com/zulip/zulip/pull/29937. 

Maybe you can chat with @sahil839, but I think you might be able to build this on top of #29937, which should merge in the next couple days.

roanster007 commented:
>So I guess I should have made sure folks were aware of this previously, but all future permissions settings need to be done using groups (see https://github.com/zulip/zulip/issues/19525), and likely building on the approach in https://github.com/zulip/zulip/pull/29937.
Maybe you can chat with sahil839, but I think you might be able to build this on top of https://github.com/zulip/zulip/pull/29937, which should merge in the next couple days.

thanks for letting me know! I'll update it.

roanster007 commented on zerver/models/realms.py:
```diff
@@ -274,6 +281,13 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+    resolve_topic_group = models.ForeignKey(
+        "UserGroup", on_delete=models.RESTRICT, related_name="+", null=True, blank=True
+    )
+
```
I'm not sure if `blank=True` is needed.

roanster007 commented on web/src/server_events_dispatch.js:
```diff
@@ -302,7 +303,7 @@ export function dispatch_normal_event(event) {
-                                if (key === "edit_topic_policy") {
+                                if (key === "edit_topic_policy" || key === "resolve_topic_group") {
                                     message_live_update.rerender_messages_view();
                                 }
```
so as to update the resolve topic icon in the message list.

timabbott replied:
This is probably worth a brief comment here for this block of the form:

> // Technically we just need to rerender the message recipient bars to update the buttons for editing or resolving a topic, but because these policies are changed rarely, it's fine to rerender the entire message feed.

roanster007 commented on web/src/user_groups.ts:
```diff
@@ -240,6 +240,10 @@ export function get_realm_user_groups_for_dropdown_list_widget(
+    if (setting_name === "resolve_topic_group") {
+        group_setting_config = {...group_setting_config, allow_nobody_group: false};
+    }
+
```
This is done to remove the "role:nobody" from dropdown options in case of `resolve_topic_group` since it is represented by checkbox.

timabbott replied:
I think it could be OK to just offer the nobody group in the set of options, and have picking that value in the menu end up toggling the checkbox in the UI.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -824,6 +842,43 @@ export function init_dropdown_widgets() {
+    if (
+        realm.realm_resolve_topic_group === user_groups.get_user_group_from_name("role:nobody").id
+    ) {
+        update_resolve_topic_sub_settings(false);
```
When initial group of the `resolve_topic_group` is "role:nobody", we don't want to show it in dropdown option. Hence, we rather show the checkbox as unchecked and dropdown with "role:members" (default).

roanster007 replied:
>We can just show the option which was selected before the setting was disabled and I guess this is what we do for other similar settings.

The problem here is that we are using just one database field for both `allow_resolve_topic` and `can_resolve_topics_group` settings. So technically, if you uncheck the checkbox, the value of this field is `role:nobody` -- which removes the possibility of storing the last option of dropdown. Hence, I thought I would just simply show the default value (members).

roanster007 commented on web/src/settings_org.js:
```diff
@@ -1109,6 +1175,11 @@ export function build_page() {
+        const is_checked = $(e.target).prop("checked");
+        update_resolve_topic_sub_settings(is_checked);
+    });
+
```
This is for disabling/enabling the dropdown depending on the `allow_resolve_topic` checkbox.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -506,6 +510,16 @@ export function discard_property_element_changes(elem, for_realm_default_setting
+                    property_name,
+                    property_value,
+                );
+            } else {
```
When the permission is set to `role:nobody` using checkbox, when `resolve_topic_group` is updated, we set the dropdown value with updated one only if the value !== `role:nobody`.id. Else, we just disable dropdown.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -874,6 +929,17 @@ export function populate_data_for_request(
+                    if (input_value === false) {
+                        user_group = user_groups.get_user_group_from_name("role:nobody");
+                        data.resolve_topic_group = user_group.id;
+                    } else {
```
When submitting request, if `allow_resolve_topic` checkbox is checked, we send a request of `resolve_topic_group = dropdown value`. Else, we send request of `resolve_topic_group` = `role:nobody`.id.

roanster007 commented on zerver/views/message_edit.py:
```diff
@@ -93,6 +94,30 @@ def fill_edit_history_entries(
+def is_resolve_topic_request(
+    user_profile: UserProfile,
+    message_id: PathOnly[NonNegativeInt],
+    new_topic_name: OptionalTopic,
```
This is used for determining if the message edit request is a resolve topic request, and calling the corresponding resolve topic method in actions. Not sure if `is_resolve_topic_request` is a good name.

roanster007 commented on zerver/views/message_edit.py:
```diff
@@ -93,6 +94,30 @@ def fill_edit_history_entries(
+        ) and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+
+        if topic_resolved or topic_unresolved:
+            return True
```
I believe the conditions are enough for determining if it is a resolve topic request?

roanster007 commented:
[review](https://chat.zulip.org/#narrow/stream/378-api-design/topic/managing.20resolve.20topic.20permissions.20independently/near/1796685).
>I think can_resolve_topics_group would be a name that fits the pattern we've begun to use in other places (like can_remove_subscribers_group).

Renamed `resolve_topic_group` to `can_resolve_topics_group`.

roanster007 commented on zerver/views/realm.py:
```diff
@@ -60,6 +63,30 @@ def parse_jitsi_server_url(
+        return setting_value
+
+    assert len(setting_value["direct_members"]) == 0 and len(setting_value["direct_subgroups"]) == 1
+
```
`views/realm` doesn't yet have a setting which  would return `AnonymousSettingGroupDict`. So to get coverage I removed that part.

Maybe, a better way would have been to add `no coverage` ?

roanster007 commented on zerver/views/realm.py:
```diff
@@ -60,6 +63,34 @@ def parse_jitsi_server_url(
+        return setting_value["direct_subgroups"][0]
+
+    return AnonymousSettingGroupDict(  # nocoverage
+        direct_members=setting_value["direct_members"],
```
No setting that returns `AnonymousSettingGroupDict` currently in `views/realm`. Hence, `no coverage` for now

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1363,3 +1363,61 @@ def check_update_message(
+    and raises a JsonableError if otherwise.
+    It returns the number changed.
+    """
+    message = access_message(user_profile, message_id, lock_message=True)
```
This is also called in `is_resolve_topic_request`, I guess we should try removing one of them.

roanster007 replied:
updated!

sahil839 commented on zerver/views/realm.py:
```diff
@@ -60,6 +63,34 @@ def parse_jitsi_server_url(
+            ]
+        ),
+    ]
+)
```
I was working on a PR for converting this to `typed_endpoint`. I will try to open the PR tommorrow, so that we do not need these changes.

roanster007 commented on zerver/views/realm.py:
```diff
@@ -101,6 +132,9 @@ def update_realm(
     edit_topic_policy: Optional[int] = REQ(
         json_validator=check_int_in(Realm.EDIT_TOPIC_POLICY_TYPES), default=None
     ),
+    can_resolve_topics_group_id: Optional[Union[Dict[str, List[int]], int]] = REQ(
```
Ya, I just realised that there is an assertion that is preventing me to access the group setting if `permission_configuration.require_system_group` is True, which in this case would hence raise error.

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1363,3 +1363,61 @@ def check_update_message(
+        raise JsonableError(_("Direct messages cannot be moved to channels."))
+
+    if not user_profile.can_resolve_topic():
+        raise JsonableError(_("You don't have permission to resolve topics."))
```
Only the `message.is_stream_message()` is duplicated.  `user_profile.can_resolve_topic()` is a new method. So I guess it is fine?

roanster007 replied:
the problem with calling it from `check_update_message` is that there are also lots of checks for the time limit for editing, but according to this issue - `There should not be an option to time-limit topic resolution.`.

Another problem is it only checks the stream post policy if `stream_id` is None. For resolving the topics, we also need to check it here.

sahil839 replied:
Maybe we can do something like - extract all the topic editing permissions in a function and call it only when topic is being edited but not resolved and have a separate function which checks all the conditions of resolving a topic. This way we would be able to do all this in a single function and also keep the code readable.

roanster007 replied:
extracted in the first commit.

roanster007 commented on zerver/models/realms.py:
```diff
@@ -657,6 +662,22 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+            allow_everyone_group=False,
+            default_group_name=SystemGroups.MEMBERS,
+            id_field_name="can_resolve_topics_group_id",
+            allowed_system_groups=[
```
updated!

roanster007 commented:
@sahil839 addressed your reviews.

sahil839 reviewed (commented):
Posted some comments.

Much of the frontend code would have been simpler if could just avoid the checkbox and keep the dropdown with "Nobody" option, but yeah having a checkbox to disable resolving seems good from UX perspective so it is fine for now.

roanster007 commented on web/src/message_list_view.js:
```diff
@@ -183,6 +181,16 @@ function set_topic_edit_properties(group, message) {
+    }
+    const sub = stream_data.get_sub(stream_name);
+    group.user_can_resolve_topic =
+        stream_data.can_post_messages_in_stream(sub) && settings_data.user_can_resolve_topic();
```
yup, it is related to this issue - https://github.com/zulip/zulip/issues/21281

sahil839 replied:
Yeah it is related, but #21811 mentions that the behavior should be discussed.

roanster007 commented on web/src/settings_components.ts:
```diff
@@ -891,6 +909,9 @@ export function check_property_changed(
             proposed_val = get_input_element_value(elem, "field-data-setting");
             break;
+        case "realm_allow_resolve_topic":
+            proposed_val = get_input_element_value(elem, "checkbox");
```
oh ya right, thanks for catching it. Must have left it when I updated the setting.

roanster007 commented on web/src/settings_components.ts:
```diff
@@ -1000,6 +1021,30 @@ export function populate_data_for_request(
+                    const group_id = can_resolve_topics_group_widget.value();
+                    assert(typeof group_id === "number");
+                    data.can_resolve_topics_group = group_id;
+                    continue;
```
ya you're right. it makes it a lot cleaner.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -824,6 +846,44 @@ export function init_dropdown_widgets() {
+        unique_id_type: dropdown_widget.DataTypes.NUMBER,
+    });
+    settings_components.set_can_resolve_topics_group_widget(can_resolve_topics_group_widget);
+    can_resolve_topics_group_widget.setup();
```
sure! do you want me to update in this rather?

sahil839 replied:
This is merged in main, so you can rebase the PR and update this.

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1363,3 +1363,59 @@ def check_update_message(
+        if not user_profile.can_move_messages_between_streams():
+            raise JsonableError(_("You don't have permission to move this message"))
+        new_stream = access_stream_by_id(user_profile, stream_id, require_active=True)[0]
+        check_stream_access_based_on_stream_post_policy(user_profile, new_stream)
```
ya it checks the permission in the original stream in the else statement below.

roanster007 replied:
As per Tim's comment [here](https://chat.zulip.org/#narrow/channel/101-design/topic/managing.20resolve.20topic.20permissions.20independently.20.2321811/near/1838294):
>Can we just leave interactions with stream posting policies follow-up project, to manage change complexity?

stream post policy is out of picture here.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -567,6 +585,10 @@ export function discard_property_element_changes(elem, for_realm_default_setting
             break;
+        case "allow_resolve_topic":
+            settings_components.set_input_element_value($elem, property_value);
+            update_resolve_topic_sub_settings(property_value);
```
ya you're right. updated!

roanster007 commented:
addressed all the reviews. posted replies for [this](https://github.com/zulip/zulip/pull/29965#discussion_r1611449173) and [this](https://github.com/zulip/zulip/pull/29965#discussion_r1613338933)

roanster007 commented on zerver/views/realm.py:
```diff
@@ -357,20 +365,35 @@ def update_realm(
-            setting_group_id_name
-        ] != getattr(realm, setting_group_id_name):
-            user_group_id = req_group_setting_vars[setting_group_id_name]
+        setting_value = req_group_setting_vars[setting_group_id_name]
+
+        if setting_value is None:
+            continue
+
+        new_setting = getattr(setting_value, "new", setting_value)
+        new_setting_value = parse_group_setting_value(new_setting)
+
+        expected_current_setting_value = None
+        old_setting = getattr(setting_value, "old", None)
+        if old_setting is not None:
+            expected_current_setting_value = parse_group_setting_value(old_setting)
+
+        current_value = getattr(realm, setting_group_id_name)
+        if validate_group_setting_value_change(
+            current_value, new_setting_value, expected_current_setting_value
+        ):
+            setting_user_group = new_setting_value
```
I had to make bunch of changes here. Do you expect them to be in a separate prep commit?

roanster007 commented:
Updated to use `GroupSettingChangeRequest`

roanster007 commented:
@sahil839 this was updated to use `GroupSettingChangeRequest`. I guess it is better to wait for your PR #30301 to get merged so that `views/realm` would be prepared to be able to receive `GroupSettingChangeRequest` settings?

sahil839 commented:
Yeah that might be better. That PR contains some test changes as well which can be just used here.

roanster007 commented:
Updated!

roanster007 commented on zerver/tests/test_realm.py:
```diff
@@ -1594,7 +1595,13 @@ def do_test_realm_permission_group_setting_update_api_with_anonymous_groups(
         othello = self.example_user("othello")
         hamlet = self.example_user("hamlet")
-        leadership_group = check_add_user_group(realm, "leadership", [hamlet], acting_user=None)
+
+        try:
+            leadership_group = check_add_user_group(realm, "leadership", [hamlet], acting_user=None)
+        except JsonableError:
+            leadership_group = NamedUserGroup.objects.get(
+                name="leadership", realm=realm, is_system_group=False
+            )
```
when the method is being called by multiple settings, then `check_add_user_group` would throw error since group already present in db.

roanster007 replied:
updated!

roanster007 commented:
@sahil839 could you have another look at this.

sahil839 commented on web/src/settings_components.ts:
```diff
@@ -1016,6 +1035,28 @@ export function populate_data_for_realm_settings_request(
+                    data.can_resolve_topics_group = JSON.stringify({
+                        new: new_user_group_id,
+                        old: realm.realm_can_resolve_topics_group,
+                    });
```
I wonder if we can do something simple here. Using two elements for one setting field makes it somewhat non-trivial. I will think more if we can do something and will post here if I can.

sahil839 commented on web/src/user_groups.ts:
```diff
@@ -240,6 +240,14 @@ export function get_realm_user_groups_for_dropdown_list_widget(
+        group_setting_config = {
+            ...group_setting_config,
+            allow_nobody_group: false,
+            require_system_group: false,
```
I guess we do not need this, we can just allow user defined groups as well.

sahil839 reviewed (commented):
Posted some comments.

roanster007 commented on zerver/tests/test_message_move_topic.py:
```diff
@@ -1760,3 +1765,69 @@ def test_send_resolve_topic_notification_with_no_topic_messages(self) -> None:
+            admin_user,
+            id1,
+        )
+        self.assert_json_success(result)
```
done!

roanster007 commented:
Updates:
- Addressed above reviews.
- Extracted message/topic edit permissions check into a separate method so that resolve topic permissions could be called from `check_update_message` as per the [comment](https://github.com/zulip/zulip/pull/29965#discussion_r1636914871).
- Removed checkbox - `can_resolve_topic` as per the [#design](https://chat.zulip.org/#narrow/stream/101-design/topic/managing.20resolve.20topic.20permissions.20independently.20.2321811/near/1824908) thread.

roanster007 commented:
@sahil839 this is ready for another review. I mentioned the updates in the above comment.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1340,6 +1347,49 @@ def check_update_message(
             mentioned_group_ids = list(rendering_result.mentions_user_group_ids)
             check_user_group_mention_allowed(user_profile, mentioned_group_ids)
```
I think would be good if we can just keep this code to check content edit permissions in the `check_update_message` only, having this function returns `mentioned_group_ids`, etc. and other multiple fields makes it hard to read.

Though we do not allow it in the UI, but I imagine there can be a API request for editing content and also resolving topic, so we should make sure handle it correctly if we allow such requests.

sahil839 reviewed (commented):
Posted some more comments.

roanster007 commented on web/src/settings_org.js:
```diff
@@ -492,6 +492,12 @@ export function discard_realm_property_element_changes(elem) {
                 property_value,
             );
             break;
+        case "realm_can_resolve_topics_group":
```
oh ya, my bad

roanster007 commented on zerver/tests/test_event_system.py:
```diff
@@ -1170,7 +1170,7 @@ def test_queries(self) -> None:
         realm = get_realm_with_settings(realm_id=user.realm_id)
 
-        with self.assert_database_query_count(44):
+        with self.assert_database_query_count(46):
```
ya you're right. thanks for mentioning.

roanster007 commented on zerver/tests/test_realm.py:
```diff
@@ -1589,12 +1589,11 @@ def do_test_realm_permission_group_setting_update_api(self, setting_name: str) -
 
     def do_test_realm_permission_group_setting_update_api_with_anonymous_groups(
-        self, setting_name: str
+        self, setting_name: str, anonymous_group: NamedUserGroup
```
done

roanster007 commented on web/src/user_groups.ts:
```diff
@@ -278,6 +292,12 @@ export function get_realm_user_groups_for_dropdown_list_widget(
                 throw new Error(`Unknown group name: ${group.name}`);
             }
+
+            group.display_name = maybe_override_group_display_name_for_dropdown(
+                setting_name,
+                group,
+            );
+
```
this section is used to replace `Nobody` with `Disable resolving topics` in dropdown. Is it fine, or you expect something better?

roanster007 commented:
@sahil839 thanks, I addressed the comments!

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1145,6 +1145,28 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+) -> bool:
+    old_topic_name = message.topic_name()
+
+    if content is None and new_topic_name is not None and propagate_mode == "change_all":
```
And also same thing regarding the `propagate_mode`, did we allow any other `propagate_mode` value while editing the topic to be prefixed by checkmark.

roanster007 replied:
>I don't think there was this restriction of content should be none while resolving the topics before. So, am not sure we should add this here

content refers to if the `message_id` of the `check_update_message` has any content updated. While resolving topics we definitely won't update this content, we just update the message content. And resolve topic requests always move all messages to new topic -- hence the `change_all`

this is the request being passed for resolving topics in web:
```js
    const request = {
        propagate_mode: "change_all",
        topic: new_topic_name,
        send_notification_to_old_thread: false,
        send_notification_to_new_thread: true,
    };
 ```

sahil839 replied:
But any request can be made by directly calling the API using something like `curl` and there was no restriction before to not allow changing content when resolving topics.

roanster007 replied:
updated!

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1356,6 +1378,20 @@ def get_message_update_data(
+    stream_id: Optional[int],
+) -> None:
+    if not user_profile.can_resolve_topic():
+        raise JsonableError(_("You don't have permission to resolve/unresolve topics."))
```
We don't have any checks for whether it is a stream message or not. I think if we want to allow editing content and `propagate_mode` values other than `change_all` as mentioned in one of the other comments, then we can just reuse `validate_message_edit_payload` function

roanster007 replied:
its checked in `is_resolve_topic_request`

sahil839 replied:
I cannot see any such check in `is_resolve_topic_request`.

roanster007 replied:
updated to reuse `validate_message_edit_payload`

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1356,6 +1378,20 @@ def get_message_update_data(
+    if stream_id is None:
+        stream_id = message.recipient.type_id
+        stream = get_stream_by_id_in_realm(stream_id, user_profile.realm)
+        check_stream_access_based_on_stream_post_policy(user_profile, stream)
```
I am not sure whether we want to check the permission is old stream even when the topic is being moved to new stream.

roanster007 replied:
`stream_id` is None here indicates that message is not being moved to new stream. So, `stream_post_policy` of current stream is being checked when the topic is resolved in this case.

sahil839 replied:
I was thinking whether we need to check posting permission in both old and new streams. Anyways, I just started a discussion on https://chat.zulip.org/#narrow/stream/101-design/topic/managing.20resolve.20topic.20permissions.20independently.20.2321811 for clarifying the behavior with respect to posting permissions.

roanster007 replied:
thanks! I should have posted there earlier -- would've avoided avoided the confusion.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1409,9 +1448,10 @@ def check_update_message(
-        message, user_profile, stream_id, topic_name, propagate_mode, edit_limit_buffer
-    )
+    if not is_resolve_topic_request(message, topic_name, propagate_mode, content):
+        check_time_limit_for_stream_or_topic_changes(
```
I guess this will not check the stream moving time limit even when stream is being changed if the topic is being updated by adding the checkmark as prefix.

roanster007 replied:
According to the issue description,
> There should not be an option to time-limit topic resolution.

So, I'm not sure whether that check should be performed.

sahil839 replied:
But if the stream is being changed, then time limit for changing stream should be checked.

roanster007 replied:
updated!

alya commented:
> One thing that I am not really sure about is how to handle the cases where multiple things are being changed, like along with topic being changed to be a resolved topic, stream or content is also changed.

I guess the user should have all the permissions in order to do that... We probably just want to not change anything (giving an error) if part of the change is not allowed.

> Also, do we want to allow partially resolving the topics. Of course one can do that by just editing topic but do we want to consider it as resolving a topic and check the permission using the new setting.

Is the question about, e.g., someone taking messages from a resolved topic and moving them into another topic without the checkmark (or, conversely, pasting in a checkmark)? Hmm. Maybe we should allow that as long as there are some messages remaining in the original topic. That way if someone doesn't think the topic should have been resolved/unresolved, they can still track that down.

alya commented:
In practice, I expect the "You can resolve topics, but can't rename them" configuration to be much more common than the other way around.

alya commented:
We can discuss more on CZO if any of these decisions are important to get right / hard to undo for the technical implementation.

sahil839 commented:
> Is the question about, e.g., someone taking messages from a resolved topic and moving them into another topic without the checkmark (or, conversely, pasting in a checkmark)? Hmm. Maybe we should allow that as long as there are some messages remaining in the original topic. That way if someone doesn't think the topic should have been resolved/unresolved, they can still track that down.

What I meant was when using the UI for resolving topics like using the checkmark in the message header or the "Resolve topic" option from popover, we always resolve the whole topic. But, if user just adds/removes a checkmark from topic edit UI and applies the edit to a part of the topic, should we only check the resolve topic permissions or the topic edit permissions like time limit as well.

roanster007 commented on zerver/tests/test_message_move_topic.py:
```diff
@@ -1760,3 +1766,124 @@ def test_send_resolve_topic_notification_with_no_topic_messages(self) -> None:
+            cordelia,
+            id3,
+        )
+        self.assert_json_success(result)
```
it already contains `staff_group`

sahil839 replied:
I meant that we should also check whether `admin_user` is allowed when the setting is set to `cordelia and staff` and `admin_user` is member of `staff_group`.

roanster007 replied:
updated!

roanster007 commented on zerver/tests/test_event_system.py:
```diff
@@ -1170,7 +1170,7 @@ def test_queries(self) -> None:
         realm = get_realm_with_settings(realm_id=user.realm_id)
 
-        with self.assert_database_query_count(44):
+        with self.assert_database_query_count(45):
```
oh, ya right

roanster007 commented:
@sahil839 removed `stream_post_policy` changes from this PR, and addressed the remaining reviews.

sahil839 reviewed (commented):
I have not done a complete review. But did reviewed some part of the code few days ago. Posting those comments for now. Will do a complete review once I get some release related work off my plate, probably in the next week.

roanster007 commented on web/tests/user_groups.test.js:
```diff
@@ -355,6 +355,16 @@ run_test("get_realm_user_groups_for_dropdown_list_widget", () => {
+                allow_everyone_group: false,
+                default_group_name: "role:members",
+                id_field_name: "can_resolve_topics_group_id",
+                allowed_system_groups: [],
```
yup, updated!

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1237,6 +1237,66 @@ def check_time_limit_for_change_all_propagate_mode(
+    edit_limit_buffer: int,
+) -> None:
+    if content is not None:
+        validate_user_can_edit_message(user_profile, message, edit_limit_buffer)
```
updated!

roanster007 replied:
makes sense!

roanster007 commented on web/tests/settings_data.test.js:
```diff
@@ -371,6 +371,51 @@ test_realm_group_settings(
+
+    realm.realm_can_resolve_topics_group = 3;
+    current_user.user_id = everyone;
+    assert.equal(settings_data.user_can_resolve_topic(), false);
```
yup, updated!

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1145,6 +1145,28 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+    message: Message,
+    new_topic_name: Optional[str],
+    propagate_mode: str,
+    content: Optional[str],
```
updated!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4533,6 +4533,19 @@ paths:
+                                            of a stream.
+
+                                            This setting cannot be set to `"role:internet"` and `"role:owners"`
+                                            system groups.
```
This line is not required -- it's both incorrect in this case given the configuration above, but also the "group-setting-values" page documents how to determine based on data the server provides which values are valid (without needing to worry about changes to it appearing only in docs); so we actively don't want lines like this anymore.

roanster007 replied:
sure, removed it!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8310,7 +8324,12 @@ paths:
 
-        **Changes**: In Zulip 10.0 (feature level 316), `edit_topic_policy`
+        **Changes**: Prior to Zulip 10.0 (feature level 319), the permission to
+        resolving a topic was managed by `can_move_messages_between_topics_group`.
```
"permission for resolving" would be better grammar.

roanster007 replied:
oops, ya!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4533,6 +4533,19 @@ paths:
+                                            system groups.
+
+                                            **Changes**: New in Zulip 10.0 (feature fevel 319). Previously, the
+                                            resolve topics permission was managed by `can_move_messages_between_topics_group`.
```
I think this phrasing would be better written as "Previously, permission to resolve topics was controlled by the more general `can_move_messages_between_topics_group` permission for moving messages."

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1152,6 +1152,28 @@ def user_info(um: UserMessage) -> dict[str, Any]:
+        if topic_resolved or topic_unresolved:
+            return True
+
+    return False
```
This check doesn't seem right -- or at least, it seems potentially wrong? I think we need to be checking that the topic is identical except for the difference in `RESOLVED_TOPIC_PREFIX`, except slightly more complex thanks to the issue of truncation of very long topic names...

timabbott replied:
Surely we have other logic in this file that would be more suitably reused for this check... yeah why don't you try extracting for reuse the first part of these functions?
```
def maybe_send_resolve_topic_notifications(                                                             
    *,                                                                                                  
    user_profile: UserProfile,                                                                          
    stream: Stream,                                                                                     
    old_topic_name: str,                                                                                
    new_topic_name: str,                                                                                
    changed_messages: QuerySet[Message],                                                                
    pre_truncation_new_topic_name: str,                                                                 
) -> tuple[int | None, bool]:                                                                           
    """Returns resolved_topic_message_id if resolve topic notifications were in fact sent."""           
    # Note that topics will have already been stripped in check_update_message.                         
    resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)                                                    
    topic_resolved: bool = (                                                                            
        new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                                
        and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                        
        and pre_truncation_new_topic_name[resolved_prefix_len:] == old_topic_name                       
    )                                                                                                   
    topic_unresolved: bool = (                                                                          
        old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                                
        and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                        
        and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == new_topic_name                              
    )                                                                                                   
                                                                                                        

```

timabbott replied:
Maybe some other reorganization is appropriate -- I feel like structurally, we want to just be computing these values once at the start and using those values throughout, not parsing it like 5 times. 

Possibly we should start with a preparatory PR refactoring this code path more deeply, perhaps using a new dataclass (or collection of 3 dataclasses) similar in nature to `MessageSendRequest` that can contain computed state like this and be passed around. Like, a `ResolveTopicRequest` object could have these values, an `MessageContentEditRequest` can have the stuff for editing content only, and a `MoveMessagesRequest` can contain the state ... and we can try to arrange that an initial set of logic does the parsing/checking and generates one of those 3 types, which can then be used with `isinstance` checks and the like to do the rest of the message-edit flow.

Would you be up for attempting some version of that? I feel like structurally, this function is getting unwieldy and breaking it up would be nice, but these tiny helpers with so many parameters don't feel like a nice way to read something like this either.

roanster007 replied:
ya, I agree this code path doesn't actually look so good. Let me give that a try!

timabbott replied:
Was the plan for this refactor to be a little prep PR?

sahil839 replied:
Opened #32845.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1244,6 +1266,71 @@ def check_time_limit_for_change_all_propagate_mode(
+    message: Message,
+    stream_id: int | None,
+) -> None:
+    if not user_profile.can_resolve_topic():
```
I'm not sure this function is a useful helper; leaving it in the caller seems like it might be more readable?

roanster007 replied:
that seems reasonable... it would've made sense previously when the method had some other things as well

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1244,6 +1266,71 @@ def check_time_limit_for_change_all_propagate_mode(
+        check_time_limit_for_change_all_propagate_mode(message, user_profile, topic_name, stream_id)
+
+
+def check_message_or_topic_edit_permissions(
```
I think this should be called `check_move_messages_within_channel_permissions` -- that at least seems to be the only things it's checking.

But I'm also not sure it's net helpful to have this function exist at all -- maybe the better boundary would be to just have a helper `check_move_messages_time_limit`, and have the first `user_profile.can_move_messages_to_another_topic` check live in the caller.

roanster007 replied:
>have a helper check_move_messages_time_limit, and have the first user_profile.can_move_messages_to_another_topic check live in the caller.

This definitely sounds better

sahil839 commented on web/src/user_groups.ts:
```diff
@@ -430,7 +430,9 @@ export function is_user_in_setting_group(
-    if (setting_name === "direct_message_permission_group" && name === "Nobody") {
+    if (setting_name === "can_resolve_topics_group" && name === "Nobody") {
+        return $t({defaultMessage: "Disable resolving topics"});
+    } else if (setting_name === "direct_message_permission_group" && name === "Nobody") {
```
I think this function is no longer needed since we use the pills UI for both of these settings. This function can be removed in a preparatory commit.

timabbott replied:
Hmm, yeah, that seems right to me.

alya commented:
@roanster007 will you have time to address the feedback here in the next few days? If not, it might be best for @sahil839 to take over getting this feature finished.

roanster007 commented:
Sorry for the delay, give me 1-2 days, if I am not able to address all of them, I'll let you know!

alya commented:
Sounds like a plan, thanks!

roanster007 commented:
@sahil839 I don't think I would be able to continue work on this at least for 2 days. Please feel free to pick it up in case it is a priority for the release.

Unaddressed reviews include:
- Preparatory commit for https://github.com/zulip/zulip/pull/29965#discussion_r1833538811
- Preparatory commit for https://github.com/zulip/zulip/pull/29965#discussion_r1846150650
- https://github.com/zulip/zulip/pull/29965#discussion_r1833535530 not sure about this

timabbott commented:
@sahil839 can you review this one?

timabbott commented on zerver/migrations/0650_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,35 @@
+    Realm.objects.filter(can_resolve_topics_group=None).update(
+        can_resolve_topics_group=NamedUserGroup.objects.filter(
+            name=MEMBERS_GROUP_NAME, realm=OuterRef("id"), is_system_group=True
+        ).values("pk")
```
Remind me where we proposed this default -- did we not want to default it to exactly whatever the "move topics" permission was previously?

roanster007 replied:
We didn't discuss anything about it previously, but yeah defaulting it to `can_move_messages_between_topics_group` does make sense, I'll update it

timabbott commented on zerver/models/realms.py:
```diff
@@ -766,6 +771,13 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+            require_system_group=False,
+            allow_internet_group=False,
+            allow_nobody_group=True,
+            allow_everyone_group=False,
```
I think `allow_everyone_group` should probably be `True` here; it's reasonable to want to allow guests to resolve topics.

roanster007 replied:
updated!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16653,6 +16670,17 @@ paths:
+                          - description: |
+                              A [group-setting value](/api/group-setting-values) defining
+                              the set of users who have permission to resolve topics
+                              of a stream.
```
"who have permission to [resolve topics](/help/resolve-a-topic) in the organization" is probably clearer.

roanster007 replied:
updated!

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1340,9 +1364,13 @@ def check_update_message(
-        )
+        if is_resolve_topic_request(message, topic_name, propagate_mode, content):
+            if not user_profile.can_resolve_topic():
+                raise JsonableError(_("You don't have permission to resolve/unresolve topics."))
```
Maybe post about this in chat.zulip.org for approval, but I think "You don't have permission to resolve topics." would be better. I think users will figure out the "unresolve" case just fine, and listing both can feel confusing and challenging for translators.

(The other option would be to just have two strings, one for "permission to resolve" and another for "permission to unresolve", but I think given that we expect UI in major clients to just hide the menu option, and thus this error won't be seen often, I don't think that investment is worth it).

alya replied:
@timabbott 's suggestion sounds good to me.

roanster007 replied:
Updated it to use "You don't have permission to resolve topics."

timabbott commented:
Did a quick review pass, since I'm still hoping to include this one in 10.0.

roanster007 commented:
@timabbott this is ready for another review!

timabbott commented:
We merged @sahil839's big refactoring commit that should make this a lot more accessible. It looks like this is rebased onto that, which is great; I'll try to do a quick review.

timabbott commented on zerver/lib/types.py:
```diff
@@ -389,3 +387,24 @@ class DirectMessageEditRequest:
+    topic_unresolved: bool
+    target_topic_name: str
+    orig_topic_name: str
+    orig_stream: "Stream"
```
Probably worth adding a comment here explaining that due to MAX_TOPIC_LENGTH, `target_topic_name` does not always different from orig_topic_name only in `RESOLVED_TOPIC_PREFIX`, just to explain the types here.

roanster007 replied:
updated!

timabbott commented on zerver/migrations/0657_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,30 @@
+
+    Realm.objects.filter(can_resolve_topics_group=None).update(
+        can_resolve_topics_group=F("can_move_messages_between_topics_group")
+    )
```
I think this is incorrect/unsafe in the case that the group is not a named group -- we have an invariant that anonymous `UserGroup` objects are used for exactly 1 setting, so just copying like this would mean that changing one group changed the other in the server internals (yikes!).

timabbott replied:
So we want to do this copy for `NamedUserGroup` values, and then create a duplicate `UserGroup` for the ones where that's needed.

roanster007 replied:
oops! I believe using usergroup_ptr as in:
```
 Realm.objects.filter(can_resolve_topics_group=None).update(
        can_resolve_topics_group=NamedUserGroup.objects.filter(
            realm=OuterRef("id"), usergroup_ptr=OuterRef("can_move_messages_between_topics_group")
        ).values("pk")
    )
```
may be right?

roanster007 replied:
I guess assigning it with a `NamedUserGroup` whose `usergroup_ptr` points to that of `can_move_messages_between_topics` should make sense?

```py
Realm.objects.filter(can_resolve_topics_group=None).update(
        can_resolve_topics_group=NamedUserGroup.objects.filter(
            realm=OuterRef("id"), usergroup_ptr=OuterRef("can_move_messages_between_topics_group")
        ).values("pk")
    )
```

timabbott commented on zerver/tests/test_message_move_topic.py:
```diff
@@ -1784,3 +1785,119 @@ def test_resolve_empty_string_topic(self) -> None:
+            admin_user,
+            id3,
+        )
+        self.assert_json_success(result)
```
I feel like we may want to move away from having every permissions test check all the cases of how a group-setting value can be constructed, given how much logic they share. But that is mostly a thought for later refactoring.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -577,6 +567,9 @@ def user_info(um: UserMessage) -> dict[str, Any]:
         assert mention_data is not None
 
+        # ResolveTopicRequest has no content edited.
+        assert not isinstance(message_edit_request, ResolveTopicRequest)
```
Generally a positive assertion with the list of valid classes is what we want, rather than exclusions like this.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1304,14 +1292,34 @@ def build_message_edit_request(
+    # way to get multiple in the resolved topic. And then an
+    # administrator can delete the messages in between. We consider this
+    # to be a fundamental risk of irresponsible message deletion,
+    # not a bug with the "resolve topics" feature.
```
I'm not sure this comment is quite correct just moved verbatim; "do nothing" isn't the right phrasing, it's instead "we treat this as not a resolve-topic request and will not consider sending resolved-topic notifications."

timabbott replied:
Hmm, thinking about this more, I think this logic isn't quite right. In particular, we don't want it to be possible to cause a topic to appear resolved via this class of multi-step move:
- Move from `foo` to `{resolved_topic_prefix} bar` to `{resolved_topic_prefix} foo`.

So I think we need to separate "send a notification about resolving a topic", which we may want to limit to the case where weird stuff didn't happen from "makes a topic appear resolved/unresolved and thus needs to check the permission". So I think we want a `changes_topic_resolution_status: true` field on `StreamMessageEditRequest` requests where the old topic and new topic differ in "does this start with `RESOLVED_TOPIC_PREFIX`" (possibly with other changes too), and to be checking that in the permissions-check logic.

sahil839 replied:
Ok, I read the existing code on main and the code changes here done. I think we can keep the code much simpler, but just wanted a couple of clarifications regarding behavior -
- We send topic resolved notifications only if all the below conditions are met-
  - The new and old topic differs only by `RESOLVED_TOPIC_PREFIX` and possible truncation if adding prefix makes the topic longer than maximum allowed length.  
  - We do not edit content or stream in the same request.
  - Topic of all messages are changed, i.e. `propagate_mode == "change_all"`
- Moving and resolving topics only differed by one thing before, i.e. what type of notification is sent. And now after adding this new setting, it will also differ by what permissions to check.

sahil839 replied:
Assuming what I mentioned above is the expected behavior, a rough idea of how we should be making the changes here -
- There is no need for separate `ResolveTopicRequest`, we already have `topic_resolved` and `topic_unresolved` fields in `StreamMessageEditRequest` and I think the logic to compute it is also correct to handle weird cases like above.
```py
        resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)
        topic_resolved = (
            target_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
            and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
            and pre_truncation_target_topic_name[resolved_prefix_len:] == old_topic_name
        )
        topic_unresolved = (
            old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
            and not target_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
            and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == target_topic_name
        )
```
- We can update these fields to be `true` only if `is_content_edited == False`, `is_stream_edited == False` and `propagate_mode === "change_all"`.
- And then update the code for checking permissions and sending notifications to use these fields.

Can also push a separate commit with these changes to make it easy to review.

sahil839 replied:
Also, we previously did not had any condition of `propagate_mode`, i.e. resolve topic notification was sent even when all of the messages were not moved. So, what is the behavior that we want?

sahil839 replied:
I had a look at previous discussion regarding this [here](https://chat.zulip.org/#narrow/channel/101-design/topic/managing.20resolve.20topic.20permissions.20independently.20.2321811/near/1834113) and it says that we should check permission if only change in topic is of the prefix irrespective of whether other fields have changed or not.

So, maybe the expected behavior is that -
- Permission for resolving should be checked if the new and old topic differs only by the "checkmark" prefix and also if stream or content is changed. I am not sure about `propagate_mode` though.
- And for sending notification, maybe we can have send move message if stream is also changed otherwise send resolving topic message, even if content is also changed. Here also we should decide what to do when `propagate_mode` is not `change_all`.

timabbott replied:
Hmm, this is a bit tricky. Logically, I think moving a topic between channels and also removing `RESOLVED_TOPIC_PREFIX` or making other changes should be treated fully as a topic move, including checking those permissions, without considering the resolve-topic permissions. 

Let me discuss a bit with @alya what we want the logic to be and see if we can come up with a clear definition.

alya replied:
Chatting with @timabbott , it seems reasonable to prohibit adding/removing the resolved topic checkmark if (a) you have move permissions but not resolve permissions, and (b) no other change to the channel or topic is being made at the same time .

I have no idea if that answers all the questions; I haven't tried to grok the thread above. But that's what @timabbott asked me about. :)

sahil839 replied:
Got this, but still need some clarifications -
- If user also changes stream or content, along with adding/removing the resolved topic checkmark then what permissions should be used? And what about the notification message? Should we send move notification if stream is changed and resolved notification if content is changed.
- Also, how does adding/removing only resolved topic checkmark without any other stream or content changes when the topic being changed only for some of the messages in the topic and not all the messages affect the permissions check and notification message being sent?

sahil839 replied:
Regarding the first question -
I guess we now have different permission settings for different changes - 
- For changing content - `allow_message_editing`
- For changing stream - `can_move_messages_between_streams_group`
- Adding/removing resolved topic prefix from topic without any other change in the topic - `can_resolve_topics_group` 
- Any other change in the topic than one mentioned in the previous point - `can_move_messages_between_topics_group`

And we should just check permission for each of them separately irrespective of what other changes are being made.

sahil839 replied:
@alya @timabbott thoughts?

timabbott replied:
Sorry for the slow reply; yes, I think that is the right way to think about it.

timabbott commented:
OK I think the above is a complete review of what changes this needs before integration.

sahil839 commented on zerver/migrations/0665_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,112 @@
+    realms_to_update = []
+    anonymous_group_ids = []
+
+    with transaction.atomic():
```
I think it is important to keep transaction here to make sure either all the steps of creating `UserGroup` and adding members to them is done or neither of them, so they can be fixed by rerunning. So, do we want to do batching here?

Not sure if there will be many realms with `can_move_messages_between_topics_group` set to anonymous groups.

timabbott replied:
Let's not do batching here. This should be quite fast, as the capability to do weird things is new and thus few realms will have done them.

timabbott commented:
I think this is one of the few database migrations planned but not integrated for 10.0. 

BTW, [#api design > API feature levels rebasing idea](https://chat.zulip.org/#narrow/channel/378-api-design/topic/API.20feature.20levels.20rebasing.20idea/with/2103914) will help eliminate rebasing pain related to API feature levels.

timabbott commented on zerver/migrations/0682_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,112 @@
+            )
+
+        UserGroupMembership.objects.bulk_create(user_membership_objects)
+        GroupGroupMembership.objects.bulk_create(group_membership_objects)
```
I feel like this is a pattern we've done a few times. I wonder if it makes sense to have a helper in `zerver.lib.migrate.py` for it?

sahil839 replied:
I am not completely sure what pattern are you referring to here, but I think it is code for creating `UserGroupMembership` and `GroupGroupMembership` objects from group members and subgroups data.

Since we do not have any other setting, which uses value of existing group setting to set the default for existing realms, we do not create these membership objects anywhere else except the migrations to create system groups for existing realms. And there as well, the data that we have for memberships to create the objects from that, is in different structure than we have here, so not sure how much useful will such a helper be currently.

sahil839 commented:
Pushed again to add a comment that was suggested in previous review. Will summarize the behavior once -
- Topics are considered as being resolved or unresolved when only change in topic is to add or remove the `RESOLVED_TOPIC_PREFIX`.
- Resolved topics permissions is check irrespective of whether stream or content are changed or not.
- There is no time limit checks done for resolving/unresolving topics.
- This PR previously added code to consider the topic as being resolved or unresolved only if topics for all messages was being changed, i.e. `propagate_mode` set to `change_all`. But removed this for now since that was not the original behavior.

sahil839 commented on zerver/migrations/0682_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,112 @@
+    anonymous_group_ids = []
+
+    with transaction.atomic():
+        for realm in realms_with_setting_set_to_anonymous_group:
```
Do we need to use batching here? Not sure how large the number of realms will be with move messages between topics set to anonymous groups.

timabbott replied:
No, it's unlikely to be multiple 10Ks where it would matter.

timabbott commented on zerver/migrations/0682_set_default_value_for_can_resolve_topics_group.py:
```diff
@@ -0,0 +1,112 @@
+
+    realms_with_setting_set_to_anonymous_group = Realm.objects.filter(
+        can_move_messages_between_topics_group__named_user_group=None, can_resolve_topics_group=None
+    )
```
I think the `can_move_messages_between_topics_group__named_user_group=None` part can be skipped, since we just changed all the ones with `can_resolve_topics_group=None` and that state the line before.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1411,24 +1411,28 @@ def check_update_message(
+        if message_edit_request.topic_resolved or message_edit_request.topic_unresolved:
+            if not user_profile.can_resolve_topic():
+                raise JsonableError(_("You don't have permission to resolve topics."))
+        else:
```
I think this `else` might be wrong given how `build_message_edit_request` will set `topic_resolved` if the topic is resolved and the channel is also changed (with no other change to the topic). This needs to be resolved and tested.

timabbott replied:
Oh, no, that's handled separately, in the `is_stream_edited` block. OK.

timabbott commented:
This looks great, merged, thanks @roanster007 and @sahil839!  I'm excited to finally have this integrated; it's been quite the journey.
